File: client-clojure/src/sirqul_io_t_platform/specs/report_definition.clj
(ns sirqul-io-t-platform.specs.report-definition
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            )
  (:import (java.io File)))


(def report-definition-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :query) string?
   (ds/opt :queryName) string?
   (ds/opt :displayName) string?
   (ds/opt :columnNames) (s/coll-of string?)
   (ds/opt :nativeSql) boolean?
   (ds/opt :display) boolean?
   (ds/opt :dataSource) string?
   (ds/opt :maxLimit) int?
   (ds/opt :maxDateRange) int?
   (ds/opt :displayParams) string?
   (ds/opt :metaData) string?
   (ds/opt :columnNamesRaw) string?
   })

(def report-definition-spec
  (ds/spec
    {:name ::report-definition
     :spec report-definition-data}))

File: client-scalaz/src/main/scala/org/openapitools/client/api/ReportDefinition.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import ReportDefinition._

case class ReportDefinition (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
query: Option[String],
queryName: Option[String],
displayName: Option[String],
columnNames: Option[List[String]],
nativeSql: Option[Boolean],
display: Option[Boolean],
dataSource: Option[DataSource],
maxLimit: Option[Long],
maxDateRange: Option[Integer],
displayParams: Option[String],
metaData: Option[String],
columnNamesRaw: Option[String])

object ReportDefinition {
  import DateTimeCodecs._
  sealed trait DataSource
  case object WRITE extends DataSource
  case object ANALYTIC extends DataSource
  case object READ1 extends DataSource

  object DataSource {
    def toDataSource(s: String): Option[DataSource] = s match {
      case "WRITE" => Some(WRITE)
      case "ANALYTIC" => Some(ANALYTIC)
      case "READ1" => Some(READ1)
      case _ => None
    }

    def fromDataSource(x: DataSource): String = x match {
      case WRITE => "WRITE"
      case ANALYTIC => "ANALYTIC"
      case READ1 => "READ1"
    }
  }

  implicit val DataSourceEnumEncoder: EncodeJson[DataSource] =
    EncodeJson[DataSource](is => StringEncodeJson(DataSource.fromDataSource(is)))

  implicit val DataSourceEnumDecoder: DecodeJson[DataSource] =
    DecodeJson.optionDecoder[DataSource](n => n.string.flatMap(jStr => DataSource.toDataSource(jStr)), "DataSource failed to de-serialize")

  implicit val ReportDefinitionCodecJson: CodecJson[ReportDefinition] = CodecJson.derive[ReportDefinition]
  implicit val ReportDefinitionDecoder: EntityDecoder[ReportDefinition] = jsonOf[ReportDefinition]
  implicit val ReportDefinitionEncoder: EntityEncoder[ReportDefinition] = jsonEncoderOf[ReportDefinition]
}
